### DISCUSSION TITLE
Mutate seems to swallow errors, contrary to some documentation.

### DISCUSSION THREAD

**Discussion by zjr** [vercel/swr#547] [General] [answered]
Hello,

If I try something like the following:
```js
import { mutate } from 'swr';
…
try {
  const foo = await mutate('bar', async () => {
    throw new Error('an error occurred');
  });
  
  console.log({ foo }); // { foo: undefined }
} catch (e) {
  // this is never executed
  console.error(e);
}
```
Mutate swallows the error and it's never heard from. However, the docs seem to state that I should be able to catch this. Here's what the docs say:
```js
try {
  const user = await mutate('/api/user', updateUser(newUser))
} catch (error) {
  // Handle an error while updating the user here
}
```
That is here: https://github.com/vercel/swr#returned-data-from-mutate.

What's going on?

**Comment by zjr** [comment]
Is it supposed to be that I would do:
```js
try {
  await mutate('bar', await (async () => {
    throw new Error('an error occurred');
  })());
} catch (e) {
  console.error(e);
}
```
so that the function is run and evaluated and it's _results_ are passed to mutate?

**Comment by huozhi** [comment, marked as answer]
@zjr I've tried to run your code snippet in codesandbox, and I did recieved the error in the console. see the trace below

![image](https://user-images.githubusercontent.com/4800338/88124592-143f9900-cc00-11ea-882c-1c3dda2e355d.png)

did you try the latest version of swr (0.2.3 by the date I reply this thread)?

**Reply by zjr** [reply]
This helped me to narrow it down a bit. When you run this in an isolated JS file it works as expected, when you run it in a react component you get the behavior I'm describing.

Here is a sandbox: https://codesandbox.io/s/old-microservice-ind5c?file=/src/App.js. The error will definitely throw, but it doesn't bubble up out of mutate and so can't be caught in a try/catch.

**Reply by zjr** [reply]
PS thanks for looking into this.

**Reply by zjr** [reply]
I realize now I had a bug in my test code in this sandbox where I wasn't supplying a key (first arg) to mutate. Adding that fixed this, but still doesn't explain what's going on in my case, so, going to work on that more.

**Reply by zjr** [reply]
OK, so it appears that this happens only when you have already used `useSWR` on the key, and then use mutate. I have updated my sandbox, here's the link again: https://codesandbox.io/s/old-microservice-ind5c?file=/src/App.js. I'm sure this is something I could just be doing wrong but I can't figure what.

<img width="1309" alt="Screen Shot 2020-07-22 at 12 07 35 PM" src="https://user-images.githubusercontent.com/913633/88200278-f11ee480-cc13-11ea-8e0b-a346756ed687.png">

**Reply by huozhi** [reply]
@zjr thanks for the stable repro example. ~~I think it's related to 477 .~~ 
after looking into the like it because it enters the revalidate stage first then return before error thrown.

**Reply by zjr** [reply]
Thanks @huozhi, should I open a new issue for this? I ask because the one you've mentioned is closed.

**Reply by huozhi** [reply]
@zjr yeah I think we should have another issue to track this, I was wrong before so I added a delete line for that comment. sorry for the confusion here. 

the issue will be titled like "mutate swallow errors when using along with useSWR hook".

**Reply by zjr** [reply]
Alright, I've created that here: https://github.com/vercel/swr/issues/556

**Reply by zjr** [reply]
For posterity, this is fixed by: https://github.com/vercel/swr/pull/557.
